theano/gof/wrapper.py

```diff
@@ -0,0 +1,446 @@
+
+    def check_that_values_are_compatible(self, data, strict, allow_downcast):
+        wrap_instance = dict()
+        for i in range(self.length):
+            wrap_instance[self.fields[i]] = self.types[i].filter(getattr(data, self.fields[i]), strict, allow_downcast)
+        return Wrap(**wrap_instance)
+
+    # Returns a wrapped object with expected attributes or (in strict mode) checks that data has expected attributes.
+    def filter(self, data, strict=False, allow_downcast=None):
+        if strict:
+            try:
+                self.check_that_values_are_compatible(data, strict, allow_downcast)
+            except AttributeError as e:
+                raise TypeError('%s: strict mode: missing expected attribute in filtered data:\n%s' % (self, e))
+            except Exception as e:
+                raise TypeError('%s: strict mode: a data does not pass corresponding type filtering:\n%s' % (self, e))
+            return data
+        elif isinstance(data, dict):
+            wrap_instance = dict()
+            for i in range(self.length):
+                if self.fields[i] not in data:
+                    raise TypeError('%s: filter expects a dictionary that has attribute "%s".' % (self, self.fields[i]))
+                try:
+                    wrap_instance[self.fields[i]] = self.types[i].filter(data[self.fields[i]], strict, allow_downcast)
+                except Exception as e:
+                    raise TypeError('%s: a data does not pass filtering for attribute "%s":\n%s' % (self, self.fields[i], e))
+            return Wrap(**wrap_instance)
+        else:
+            try:
+                wrapped_data = self.check_that_values_are_compatible(data, strict, allow_downcast)
+            except AttributeError as e:
+                raise TypeError('%s: missing expected attribute in filtered data:\n%s' % (self, e))
+            except Exception as e:
+                raise TypeError('%s: a data does not pass corresponding type filtering:\n%s' % (self, e))
+            return wrapped_data
+
+    def values_eq(self, a, b):
+        for i in range(self.length):
+            if not self.types[i].values_eq(getattr(a, self.fields[i]), getattr(b, self.fields[i])):
+                return False
+        return True
+
+    def values_eq_approx(self, a, b):
+        a = self.filter(a, strict=False, allow_downcast=True)
+        b = self.filter(b, strict=False, allow_downcast=True)
+        for i in range(self.length):
+            if not self.types[i].values_eq_approx(getattr(a, self.fields[i]), getattr(b, self.fields[i])):
+                return False
+        return True
+
+    def c_compile_args(self):
+        c_compile_args_list = []
+        for _type in self.types:
+            try:
+                try:
+                    c_compile_args_list.extend(_type.c_compile_args())
+                except TypeError:
+                    c_compile_args_list.extend(_type.c_compile_args(compiler))
```

Comment:
NB: Understood, I should get the compiler from the Wrapper.c_compile_args()` args (`def c_compile_args(self, compiler)`) and pass it to underlying types when it's necessary. Same for following methods.